tcptracer-bpf.c

```diff
@@ -0,0 +1,927 @@
+
+	return bint.c[0] == 1;
+}
+
+/* check if IPs are IPv4 mapped to IPv6 ::ffff:xxxx:xxxx
+ * https://tools.ietf.org/html/rfc4291#section-2.5.5
+ * the addresses are stored in network byte order so IPv4 adddress is stored
+ * in the most significant 32 bits of part saddr_l and daddr_l.
+ * Meanwhile the end of the mask is stored in the least significant 32 bits.
+ */
+static inline bool is_ipv4_mapped_ipv6(u64 saddr_h, u64 saddr_l, u64 daddr_h, u64 daddr_l) {
+	if (is_big_endian()) {
+		return ((saddr_h == 0 && ((u32)(saddr_l >> 32) == 0x0000FFFF)) ||
+                        (daddr_h == 0 && ((u32)(daddr_l >> 32) == 0x0000FFFF)));
+	} else {
+		return ((saddr_h == 0 && ((u32)saddr_l == 0xFFFF0000)) ||
+                        (daddr_h == 0 && ((u32)daddr_l == 0xFFFF0000)));
+	}
+}
+
+#define TCPTRACER_STATUS_UNINITIALIZED 0
+#define TCPTRACER_STATUS_CHECKING      1
+#define TCPTRACER_STATUS_CHECKED       2
+#define TCPTRACER_STATUS_READY         3
+struct tcptracer_status_t {
+	u64 status;
+
+	/* checking */
+	u64 pid_tgid;
+	u64 what;
+	u64 offset_saddr;
+	u64 offset_daddr;
+	u64 offset_sport;
+	u64 offset_dport;
+	u64 offset_netns;
+	u64 offset_ino;
+	u64 offset_family;
+	u64 offset_daddr_ipv6;
+
+	u8 err;
+
+	u32 saddr;
+	u32 daddr;
+	u16 sport;
+	u16 dport;
+	u32 netns;
+	u16 family;
+	u32 daddr_ipv6[4];
+	char padding[6];
+};
+
+struct bpf_map_def SEC("maps/tcptracer_status") tcptracer_status = {
+	.type = BPF_MAP_TYPE_HASH,
+	.key_size = sizeof(__u64),
+	.value_size = sizeof(struct tcptracer_status_t),
+	.max_entries = 8,
+};
+
+static inline bool check_family(struct sock *sk, u16 expected_family) {
+	struct tcptracer_status_t *status;
+	u64 zero = 0;
+	u16 family;
+	family = 0;
+
+	status = bpf_map_lookup_elem(&tcptracer_status, &zero);
+	if (status == NULL || status->status != TCPTRACER_STATUS_READY) {
+		return 1;
+	}
+
+	bpf_probe_read(&family, sizeof(u16), ((char *)sk) + status->offset_family);
+
+	return family == expected_family;
+}
+
+SEC("kprobe/tcp_v4_connect")
+int kprobe__tcp_v4_connect(struct pt_regs *ctx)
+{
+	struct sock *sk;
+	u64 pid = bpf_get_current_pid_tgid();
+
+	sk = (struct sock *) PT_REGS_PARM1(ctx);
+
+	bpf_map_update_elem(&connectsock_ipv4, &pid, &sk, BPF_ANY);
+
+	return 0;
+}
+
+SEC("kretprobe/tcp_v4_connect")
+int kretprobe__tcp_v4_connect(struct pt_regs *ctx)
+{
+	int ret = PT_REGS_RC(ctx);
+	u64 pid = bpf_get_current_pid_tgid();
+	struct sock **skpp;
+	u64 zero = 0;
+	struct tcptracer_status_t *status;
+
+	skpp = bpf_map_lookup_elem(&connectsock_ipv4, &pid);
+	if (skpp == 0) {
+		return 0;	// missed entry
+	}
+
+	struct sock *skp = *skpp;
+
+	bpf_map_delete_elem(&connectsock_ipv4, &pid);
+
+	if (ret != 0) {
+		// failed to send SYNC packet, may not have populated
+		// socket __sk_common.{skc_rcv_saddr, ...}
+		return 0;
+	}
+
+	status = bpf_map_lookup_elem(&tcptracer_status, &zero);
+	if (status == NULL || status->status == TCPTRACER_STATUS_UNINITIALIZED) {
+		return 0;
+	}
+
+	switch (status->status) {
```

Comment:
Smart! worth commenting that we are abusing connects to figure out offsets. Also, can we factor out this part in a separate function? (assuming ebpf is OK with it)